File: frontend/src/api/services/OperationsService.ts
/* hand-written — Operations team transactions */
import type { CancelablePromise } from '../core/CancelablePromise';
import { OpenAPI } from '../core/OpenAPI';
import { request as __request } from '../core/request';

export class OperationsService {
    public static getTransactions(
        startDate?: string,
        endDate?:   string,
        limit:      number = 100,
        reasonTypes?: Array<string>,
        statuses?:    Array<string>,
        initiators?:  Array<string>,
        transactionIds?: Array<string>,
    ): CancelablePromise<Array<any>> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/transactions',
            query: {
                'start_date':      startDate,
                'end_date':        endDate,
                'limit':           limit,
                'reason_types':    reasonTypes,
                'statuses':        statuses,
                'initiators':      initiators,
                'transaction_ids': transactionIds,
            },
        });
    }

    public static getTransactionsSummary(
        startDate?: string,
        endDate?:   string,
        reasonTypes?: Array<string>,
        statuses?:    Array<string>,
        initiators?:  Array<string>,
        transactionIds?: Array<string>,
    ): CancelablePromise<any> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/transactions/summary',
            query: {
                'start_date':      startDate,
                'end_date':        endDate,
                'reason_types':    reasonTypes,
                'statuses':        statuses,
                'initiators':      initiators,
                'transaction_ids': transactionIds,
            },
        });
    }

    public static getDistinctInitiators(
        startDate?: string,
        endDate?:   string,
    ): CancelablePromise<Array<string>> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/initiators',
            query: {
                'start_date': startDate,
                'end_date':   endDate,
            },
        });
    }

    // ── E-money operations transactions (Withdrawal / Deposit) with ACCOUNT_CLASS ──
    public static getEmoneyTransactions(
        startDate?: string,
        endDate?:   string,
        limit:      number = 100,
        statuses?:        Array<string>,
        initiators?:      Array<string>,
        transactionIds?:  Array<string>,
        accountClasses?:  Array<string>,
    ): CancelablePromise<Array<any>> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/emoney-transactions',
            query: {
                'start_date':      startDate,
                'end_date':        endDate,
                'limit':           limit,
                'statuses':        statuses,
                'initiators':      initiators,
                'transaction_ids': transactionIds,
                'account_classes': accountClasses,
            },
        });
    }

    public static getEmoneyTransactionsSummary(
        startDate?: string,
        endDate?:   string,
        statuses?:        Array<string>,
        initiators?:      Array<string>,
        transactionIds?:  Array<string>,
        accountClasses?:  Array<string>,
    ): CancelablePromise<any> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/emoney-transactions/summary',
            query: {
                'start_date':      startDate,
                'end_date':        endDate,
                'statuses':        statuses,
                'initiators':      initiators,
                'transaction_ids': transactionIds,
                'account_classes': accountClasses,
            },
        });
    }

    public static getEmoneyDistinctInitiators(
        startDate?: string,
        endDate?:   string,
    ): CancelablePromise<Array<string>> {
        return __request(OpenAPI, {
            method: 'GET',
            url: '/operations/emoney-initiators',
            query: {
                'start_date': startDate,
                'end_date':   endDate,
            },
        });
    }
}

File: backend/services/operations/operations_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from repositories.operations.operations_transaction_repository import OperationsTransactionRepository
from repositories.operations.emoney_ops_transaction_repository import EmoneyOpsTransactionRepository


class OperationsService:
    @staticmethod
    def get_transactions(
        db: Session,
        start_date:   Optional[str] = None,
        end_date:     Optional[str] = None,
        limit:        int = 100,
        reason_types: Optional[List[str]] = None,
        statuses:     Optional[List[str]] = None,
        initiators:   Optional[List[str]] = None,
        transaction_ids: Optional[List[str]] = None,
    ) -> List[Dict[str, Any]]:
        return OperationsTransactionRepository(db).get_transactions(
            start_date=start_date, end_date=end_date, limit=limit,
            reason_types=reason_types, statuses=statuses,
            initiators=initiators, transaction_ids=transaction_ids,
        )

    @staticmethod
    def get_transactions_summary(
        db: Session,
        start_date:   Optional[str] = None,
        end_date:     Optional[str] = None,
        reason_types: Optional[List[str]] = None,
        statuses:     Optional[List[str]] = None,
        initiators:   Optional[List[str]] = None,
        transaction_ids: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        return OperationsTransactionRepository(db).get_summary(
            start_date=start_date, end_date=end_date,
            reason_types=reason_types, statuses=statuses,
            initiators=initiators, transaction_ids=transaction_ids,
        )

    @staticmethod
    def get_distinct_initiators(
        db: Session,
        start_date: Optional[str] = None,
        end_date:   Optional[str] = None,
    ) -> List[str]:
        return OperationsTransactionRepository(db).get_distinct_initiators(
            start_date=start_date, end_date=end_date,
        )


class EmoneyOpsService:
    @staticmethod
    def get_transactions(
        db: Session,
        start_date:   Optional[str] = None,
        end_date:     Optional[str] = None,
        limit:        int = 100,
        statuses:     Optional[List[str]] = None,
        initiators:   Optional[List[str]] = None,
        transaction_ids: Optional[List[str]] = None,
        account_classes: Optional[List[str]] = None,
    ) -> List[Dict[str, Any]]:
        return EmoneyOpsTransactionRepository(db).get_transactions(
            start_date=start_date, end_date=end_date, limit=limit,
            statuses=statuses, initiators=initiators,
            transaction_ids=transaction_ids, account_classes=account_classes,
        )

    @staticmethod
    def get_transactions_summary(
        db: Session,
        start_date:   Optional[str] = None,
        end_date:     Optional[str] = None,
        statuses:     Optional[List[str]] = None,
        initiators:   Optional[List[str]] = None,
        transaction_ids: Optional[List[str]] = None,
        account_classes: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        return EmoneyOpsTransactionRepository(db).get_summary(
            start_date=start_date, end_date=end_date,
            statuses=statuses, initiators=initiators,
            transaction_ids=transaction_ids, account_classes=account_classes,
        )

    @staticmethod
    def get_distinct_initiators(
        db: Session,
        start_date: Optional[str] = None,
        end_date:   Optional[str] = None,
    ) -> List[str]:
        return EmoneyOpsTransactionRepository(db).get_distinct_initiators(
            start_date=start_date, end_date=end_date,
        )


from core.cache import cache_all_methods  # noqa: E402
cache_all_methods(OperationsService)
cache_all_methods(EmoneyOpsService)
